tests/test_connection.py

```diff
@@ -341,3 +349,274 @@ def cb1(n):
+    assert conn.autocommit is False
+    if exception:
+        assert_rows(conn, {"prior"}, still_in_transaction=True)
+    else:
+        assert_rows(conn, {"prior", "new"}, still_in_transaction=True)
+    # Nothing committed yet; changes not visible on another connection
+    assert_rows(svcconn, set())
+
+
+def test_transaction_nested_all_changes_persisted_on_successful_exit(
+    temp_table, conn, svcconn
+):
+    """Changes from nested transaction contexts are all persisted on exit."""
+    with conn.transaction():
+        insert_row(conn, "outer-before")
+        with conn.transaction():
+            insert_row(conn, "inner")
+        insert_row(conn, "outer-after")
+    assert_rows(conn, {"outer-before", "inner", "outer-after"})
+    assert_rows(svcconn, {"outer-before", "inner", "outer-after"})
+
+
+def test_transaction_nested_all_changes_discarded_on_outer_exception(
+    temp_table, conn, svcconn
+):
+    """
+    Changes from nested transaction contexts are discarded when an exception
+    raised in outer context escapes.
+    """
+    with pytest.raises(ExpectedException):
+        with conn.transaction():
+            insert_row(conn, "outer")
+            with conn.transaction():
+                insert_row(conn, "inner")
+            raise ExpectedException()
+    assert_rows(conn, set())
+    assert_rows(svcconn, set())
+
+
+def test_transaction_nested_all_changes_discarded_on_inner_exception(
+    temp_table, conn, svcconn
+):
+    """
+    Changes from nested transaction contexts are discarded when an exception
+    raised in inner context escapes the outer context.
+    """
+    with pytest.raises(ExpectedException):
+        with conn.transaction():
+            insert_row(conn, "outer")
+            with conn.transaction():
+                insert_row(conn, "inner")
+                raise ExpectedException()
+    assert_rows(conn, set())
+    assert_rows(svcconn, set())
+
+
+def test_transaction_nested_inner_scope_exception_handled_in_outer_scope(
+    temp_table, conn, svcconn
+):
+    """
+    An exception escaping the inner transaction context causes changes made
+    within that inner context to be discarded, but the error can then be
+    handled in the outer context, allowing changes made in the outer context
+    (both before, and after, the inner context) to be successfully committed.
+    """
+    with conn.transaction():
+        insert_row(conn, "outer-before")
+        with pytest.raises(ExpectedException):
+            with conn.transaction():
+                insert_row(conn, "inner")
+                raise ExpectedException()
+        insert_row(conn, "outer-after")
+    assert_rows(conn, {"outer-before", "outer-after"})
+    assert_rows(svcconn, {"outer-before", "outer-after"})
+
+
+def insert_row(conn, value):
+    conn.cursor().execute("INSERT INTO temp_table VALUES (%s)", (value,))
+
+
+def assert_rows(conn, expected, still_in_transaction=False):
```

Comment:
I would leave `still_in_transaction` as a separate assert

Reply:
Agree. This was copied from the psycopg-nestedtransactions tests and I wondered why it was there. Will remove. Any test that cares should assert this separately.